httomo/wrappers_class.py

```diff
@@ -0,0 +1,307 @@
+from typing import Dict
+import numpy as np
+import inspect
+from mpi4py import MPI
+from mpi4py.MPI import Comm
+from inspect import signature
+from httomo.utils import print_once
+
+
+gpu_enabled = False
+try:
+    import cupy as xp
+
+    try:
+        xp.cuda.Device(0).compute_capability
+        gpu_enabled = True  # CuPy is installed and GPU is available
+    except xp.cuda.runtime.CUDARuntimeError:
+        import numpy as xp
+
+        print("CuPy is installed but GPU device inaccessible")
+except ImportError:
+    import numpy as xp
+
+    print("CuPy is not installed")
+
+
+class BaseWrapper:
+    """A parent class for all wrappers in httomo that use external modules."""
+
+    def __init__(
+        self, module_name: str, function_name: str, method_name: str, comm: Comm
+    ):
+        self.comm = comm
+        local_comm = comm.Split_type(MPI.COMM_TYPE_SHARED)
+        if gpu_enabled:
+            self.num_GPUs = xp.cuda.runtime.getDeviceCount()
+            self.gpu_id = local_comm.rank % self.num_GPUs
+            xp._default_memory_pool.free_all_blocks()
+
+    def _transfer_data(self, *args) -> tuple:
+        """transfers data between the host and device
+        Returns:
+            tuple: converted datasets
+        """
+        if not gpu_enabled:
+            return args
+        xp.cuda.Device(self.gpu_id).use()
+        if self.cupyrun:
+            return tuple(xp.asarray(d) for d in args)
+        else:
+            return tuple(xp.asnumpy(d) for d in args)
+
+    def _execute_generic(
+        self, method_name: str, params: Dict, data: xp.ndarray, reslice_ahead: bool
+    ) -> xp.ndarray:
+        """The generic wrapper to execute functions for external packages.
+
+        Args:
+            method_name (str): The name of the method to use.
+            params (Dict): A dict containing independent of httomo params.
+            data (xp.ndarray): a numpy or cupy data array.
+            reslice_ahead (bool): a bool to inform the wrapper if the reslice ahead and the conversion to numpy required.
+
+        Returns:
+            xp.ndarray: A numpy or cupy array containing processed data.
+        """
+        # set the correct GPU ID if it is required
+        if "gpu_id" in params:
+            params["gpu_id"] = self.gpu_id
+
+        # check where data needs to be transfered host <-> device
+        data = self._transfer_data(data)
+
+        data = getattr(self.module, method_name)(data[0], **params)
```

Comment:
wondering why `data[0]` here? Would that work, considering the `transfer_data` above?

Reply:
it doesn't like the `tuple` being passed...

Reply:
Doesn't `_transfer_data()` return a `tuple` of arrays (one array for each "output dataset" that a method returns I think): https://github.com/DiamondLightSource/httomo/blob/ffb0610047261a23b84090afeee0218cbdbd2d18/httomo/wrappers_class.py#L40-L51

So if `data` becomes a `tuple` of arrays at line 72:
```python
data = self._transfer_data(data)
```
then `data[0]` would be ok?

Reply:
Ah, it gets a tuple with one element now of course. Perhaps we can adjust the `transfer_data` function to keep the tuple as a local variable, say `ret`, and at the end: `return ret[0] if len(ret) == 1 else ret`

Reply:
implemented as suggested, thanks!